core/node/provider.go

```diff
@@ -5,145 +5,151 @@ import (
-				},
-				OnStop: func(ctx context.Context) error {
-					return sys.Close()
-				},
-			})
-		}
-
-		return sys
-	}
-}
-
-// BatchedProviderSys creates new provider system
-func BatchedProviderSys(isOnline bool, reprovideInterval time.Duration) interface{} {
-	return func(lc fx.Lifecycle, cr irouting.ProvideManyRouter, q *q.Queue, keyProvider simple.KeyChanFunc, repo repo.Repo) (provider.System, error) {
-		sys, err := batched.New(cr, q,
-			batched.ReproviderInterval(reprovideInterval),
-			batched.Datastore(repo.Datastore()),
-			batched.KeyProvider(keyProvider))
+func ProviderSys(reprovideInterval time.Duration) fx.Option {
+	const magicThroughputReportCount = 128
+	return fx.Provide(func(lc fx.Lifecycle, cr irouting.ProvideManyRouter, keyProvider provider.KeyChanFunc, repo repo.Repo, bs blockstore.Blockstore) (provider.System, error) {
+		sys, err := provider.New(repo.Datastore(),
+			provider.Online(cr),
+			provider.ReproviderInterval(reprovideInterval),
+			provider.KeyProvider(keyProvider),
+			provider.ThroughputReport(func(reprovide bool, complete bool, keysProvided uint, duration time.Duration) bool {
+				avgProvideSpeed := duration / time.Duration(keysProvided)
+				count := uint64(keysProvided)
+
+				if !reprovide || !complete {
+					// We don't know how many CIDs we have to provide, try to fetch it from the blockstore.
+					// But don't try for too long as this might be very expensive if you have a huge datastore.
+					ctx, cancel := context.WithTimeout(context.Background(), time.Minute*5)
+					defer cancel()
+
+					// FIXME: I want a running counter of blocks so size of blockstore can be an O(1) lookup.
+					ch, err := bs.AllKeysChan(ctx)
+					if err != nil {
+						logger.Errorf("fetching AllKeysChain in provider ThroughputReport: %v", err)
+						return false
+					}
+					count = 0
+				countLoop:
+					for {
+						select {
+						case _, ok := <-ch:
+							if !ok {
+								break countLoop
+							}
+							count++
+						case <-ctx.Done():
+							// really big blockstore mode
+
+							// how many blocks would be in a 10TiB blockstore with 128KiB blocks.
+							const probableBigBlockstore = (10 * 1024 * 1024 * 1024 * 1024) / (128 * 1024)
+							// How long per block that lasts us.
+							expectedProvideSpeed := reprovideInterval / probableBigBlockstore
+							if avgProvideSpeed > expectedProvideSpeed {
+								logger.Errorf(`
+🔔🔔🔔 YOU MAY BE FALLING BEHIND DHT REPROVIDES! 🔔🔔🔔
+
+⚠️ Your system might be struggling to keep up with DHT reprovides!
+Meaning your content being partially or completely inacessible on the network.
+We observed that you recently provided %d keys at an average rate of %v per key.
+
+🕑 An attempt to estimate your blockstore size timed out after 5 minutes,
+implying your blockstore might be exceedingly large. Assuming a considerable
+size of 10TiB, it would take %v to provide the complete set.
+
+⏰ The total provide time needs to stay under 24 hours to prevent falling behind!
```

Comment:
I thought this value was configurable and not fixed to 24 hours.  Lets use the configured value for our calculations and messaging.

Reply:
IMO this is a bug this should not be configurable.
Or at least this should be moved to one of the internal fields like the bitswap server options.

I can fix this if you want but the message will get weird trying to disambiguate the various edge cases, for example what if the user see that the estimated time is 3 days and then decide to bump the reprovide interval to 1 week ? Well the provides will happen on time but data will still be dodgly reachable since DHT pruning will happen faster than you keep up.

Reply:
I've changed this like you suggested.
The TTL for the DHT is hardcoded afaik even if I am wrong the TTL is enforced by remotes node so you can't change it anyway.
You are right danger is only if it's bigger however I don't see value in reproviding more often, you will increase your load on the network for no gain (probelab mesure churn and 24h all records are still retrievable)